### PR TITLE
Fix error on op selection

### PR DESCRIPTION
**Pull request by johannkm**
With https://github.com/dagster-io/dagster/pull/16640/files#diff-3a0d9c2e7d488c33f5be6fdf25298f9b03277b1bc73417d53e0a51688d32e8b6, we started passing assetCheckSelection: [] for op jobs. This was untested, and hit this invariant because unlike for assets where [] gets turned in to None, we maintain a difference between None and [] with asset checks

This is ripe for rethinking

### CODE REVIEW

File: python_modules/dagster/dagster/_core/instance/__init__.py [hunk 1]
```diff
@@ -1461,7 +1461,7 @@ def create_run(
         check.opt_set_param(asset_selection, "asset_selection", of_type=AssetKey)
         check.opt_set_param(asset_check_selection, "asset_check_selection", of_type=AssetCheckKey)
 
-        if asset_selection is not None or asset_check_selection is not None:
+        if asset_selection is not None or asset_check_selection:
```
Comment (gibsondan) [inline python_modules/dagster/dagster/_core/instance/__init__.py:1464]: maybe worth a comment here with historical context on why its important that these two differ?

File: python_modules/dagster-graphql/dagster_graphql/test/utils.py [hunk 2]
```diff
@@ -201,6 +201,8 @@ def infer_pipeline_selector(
         {
             "pipelineName": pipeline_name,
             "solidSelection": op_selection,
+            # "assetSelection": [],
```
Comment (sryza) [inline python_modules/dagster-graphql/dagster_graphql/test/utils.py:204]: Are these comments intentional?

File: python_modules/dagster-graphql/dagster_graphql/test/utils.py [hunk 3]
```diff
@@ -201,6 +201,8 @@ def infer_pipeline_selector(
         {
             "pipelineName": pipeline_name,
             "solidSelection": op_selection,
+            "assetSelection": [],
```
Comment (sryza) [inline python_modules/dagster-graphql/dagster_graphql/test/utils.py:204]: So this now means we have coverage for the case where these are set. Is there a case where these aren't set that we're losing coverage for?

Comment (johannkm) [reply python_modules/dagster-graphql/dagster_graphql/test/utils.py:204]: Good question. The front end always provides them now, but that's not guaranteed forever and gql api users might not. I'll add coverage

Comment (johannkm) [reply python_modules/dagster-graphql/dagster_graphql/test/utils.py]: updated with a specific case for []. Going to start a discussion about how to handle this going forwards

File: python_modules/dagster/dagster/_core/instance/__init__.py [hunk 4]
```diff
@@ -1461,7 +1461,9 @@ def create_run(
         check.opt_set_param(asset_selection, "asset_selection", of_type=AssetKey)
         check.opt_set_param(asset_check_selection, "asset_check_selection", of_type=AssetCheckKey)
 
-        if asset_selection is not None or asset_check_selection is not None:
+        # asset_selection gets coerced from [] to None, but asset_check_selection doesn't. We
+        # allow asset_check_selection to be [] when op_selection is set.
```
Comment (gibsondan) [inline python_modules/dagster/dagster/_core/instance/__init__.py:1465]: I'm still struggling to follow this. Where does the coercing being referenced here happen? Can we add some whys in addition to some whats here / be more verbose?

**Review by gibsondan — APPROVED** [review]

**Pull request merged**